src/swarm-engine/Swarm/Game/CESK.hs

```diff
@@ -295,48 +295,91 @@ instance FromJSON CESK where
 
 -- | Is the CESK machine in a final (finished) state?  If so, extract
 --   the final value and store.
-finalValue :: CESK -> Maybe (Value, Store)
+finalValue :: CESK -> Maybe Value
 {-# INLINE finalValue #-}
-finalValue (Out v s []) = Just (v, s)
+finalValue (Out v _ []) = Just v
+finalValue (Suspended v _ _ []) = Just v
 finalValue _ = Nothing
 
--- | Initialize a machine state with a starting term along with its
---   type; the term will be executed or just evaluated depending on
---   whether it has a command type or not.
-initMachine :: ProcessedTerm -> Env -> Store -> CESK
-initMachine t e s = initMachine' t e s []
-
--- | Like 'initMachine', but also take an explicit starting continuation.
-initMachine' :: ProcessedTerm -> Env -> Store -> Cont -> CESK
-initMachine' (ProcessedTerm (Module t' ctx tydefs) _ reqCtx) e s k =
-  case t' ^. sType of
-    -- If the starting term has a command type...
-    Forall _ (TyCmd _) ->
-      case (ctx, tydefs) of
-        -- ...but doesn't contain any definitions, just create a machine
-        -- that will evaluate it and then execute it.
-        (Empty, Empty) -> In t e s (FExec : k)
-        -- Or, if it does contain definitions, then load the resulting
-        -- context after executing it.
-        _ -> In t e s (FExec : FLoadEnv (Contexts ctx reqCtx tydefs) : k)
-    -- Otherwise, for a term with a non-command type, just
-    -- create a machine to evaluate it.
-    _ -> In t e s k
+-- | Extract the environment from a suspended CESK machine (/e.g./ to
+--   use for typechecking).
+suspendedEnv :: Traversal' CESK Env
+suspendedEnv = traversal go
  where
-  -- Erase all type and SrcLoc annotations from the term before
-  -- putting it in the machine state, since those are irrelevant at
-  -- runtime.
-  t = eraseS t'
+  go :: Applicative f => (Env -> f Env) -> CESK -> f CESK
+  go f (Suspended v e s k) = Suspended v <$> f e <*> pure s <*> pure k
+  go _ cesk = pure cesk
+
+-- | Lens focusing on the store of a CESK machine.
+store :: Lens' CESK Store
+store = lens get set
+ where
+  get = \case
+    In _ _ s _ -> s
+    Out _ s _ -> s
+    Up _ s _ -> s
+    Waiting _ c -> get c
+    Suspended _ _ s _ -> s
+  set cesk s = case cesk of
+    In t e _ k -> In t e s k
+    Out v _ k -> Out v s k
+    Up x _ k -> Up x s k
+    Waiting t c -> Waiting t (set c s)
+    Suspended v e _ k -> Suspended v e s k
+
+-- | Lens focusing on the continuation of a CESK machine.
+cont :: Lens' CESK Cont
+cont = lens get set
```

Comment:
Maybe... I could try.  To be honest I'm not quite sure how it works when you have a data type with multiple constructors and named fields.

Reply:
Hmm, not really, because of the `Waiting` constructor, which does not directly contain a `Store` or `Cont` field, we have to recursively get it from the nested `CESK` value.